File: src/main/java/ar/com/flow/bankaccount/Deposit.java
package ar.com.flow.bankaccount;

import ar.com.flow.bankaccount.transaction.Step;
import ar.com.flow.bankaccount.transaction.Transaction;
import ar.com.flow.bankaccount.transaction.receipt.Action;
import ar.com.flow.bankaccount.transaction.receipt.CreditPrinter;
import ar.com.flow.money.Money;

class Deposit {
    static DepositBuilder to(BankAccount creditAccount) {
        return new DepositBuilder(creditAccount);
    }

    static class DepositBuilder {
        private final BankAccount creditAccount;

        DepositBuilder(BankAccount creditAccount) {
            this.creditAccount = creditAccount;
        }

        Transaction amount(Money amountToDeposit) {
            return Transaction.builder()
                    .amount(amountToDeposit)
                    .step(new Step(creditAccount, Deposit.receipt(creditAccount)))
                    .build();
        }
    }

    static CreditPrinter receipt(BankAccount account) {
        return new CreditPrinter(account, Action.Deposit);
    }
}

File: src/main/kotlin/ar/com/flow/bankaccount/Deposit.kt
package ar.com.flow.bankaccount

import ar.com.flow.bankaccount.transaction.Builder
import ar.com.flow.bankaccount.transaction.Step
import ar.com.flow.bankaccount.transaction.Transaction
import ar.com.flow.bankaccount.transaction.receipt.Action
import ar.com.flow.bankaccount.transaction.receipt.CreditPrinter
import ar.com.flow.money.Money

internal object Deposit {
    fun to(creditAccount: BankAccount): DepositBuilder {
        return DepositBuilder(creditAccount)
    }

    fun receipt(account: BankAccount?): CreditPrinter {
        return CreditPrinter(account!!, Action.Deposit)
    }

    internal class DepositBuilder(private val creditAccount: BankAccount) {
        fun amount(amountToDeposit: Money): Transaction {
            return Builder()
                .amount(amountToDeposit)
                .step(Step(creditAccount, receipt(creditAccount)))
                .build()
        }
    }
}